parlai/core/utils.py

```diff
@@ -713,6 +739,45 @@ def __contains__(self, key):
         """Determine if text contains any offensive words in the filter."""
         return self.contains_offensive_language(key)
 
+    def str_segment(self, text, dict_agent):
+
+        @self.memo
+        def segment(text):
+            # Return a list of words that is the best segmentation of text.
+            if not text:
+                return []
+            candidates = ([first]+segment(rem) for first, rem in splits(text))
+            return max(candidates, key=prob_words)
+
+        def splits(text, L=20):
+            # Returns a list of all possible first and remainder tuples where
+            # first +" "+ rem = tex
+            candidates = []
+            for i in range(min(len(text), L)):
+                candidates.append((text[:i+1], text[i+1:]))
+            return candidates
+
+        def prob_words(words):
+            # Returns Naive Bayes probability for a sequence of words
+            return product(prob_dist(w) for w in words)
+
+        def product(nums):
+            # Return the product of a sequence of numbers
+            return reduce(operator.mul, nums, 1)
+
+        def avoid_long_words(word, N):
+            return 10./(N * 10**len(word))
+
+        N = dict_agent.maxtokens
```

Comment:
I'm not 100% certain this can be trusted, but I also didn't check the dict source.